cosmwasm/packages/wasmi-runtime/src/wasm/db.rs

```diff
@@ -56,119 +56,167 @@ pub fn read_encrypted_key(
 
     // Call read_db (this bubbles up to Tendermint via ocalls and FFI to Go code)
     // This returns the value from Tendermint
-    // fn read_db(context: Ctx, key: &[u8]) -> Option<Vec<u8>> {
-    read_db(context, &scrambled_field_name)
-        .map(|val| {
-            val.map(|as_slice| decrypt_key(&scrambled_field_name, &as_slice, contract_key))
-                .transpose()
-        })
-        .map_err(|err| {
-            error!(
-                "read_db() got an error from ocall_read_db, stopping wasm: {:?}",
-                err
-            );
-            DbError::FailedRead
-        })?
+    match read_db(context, &scrambled_field_name) {
+        Ok((value, gas_used)) => match value {
+            Some(value) => match decrypt_key(&scrambled_field_name, &value, contract_key) {
+                Ok(encrypted) => Ok((Some(encrypted), gas_used)),
+                // This error case is why we have all the matches here.
+                // If we successfully collected a value, but failed to decrypt it, then we propagate that error.
+                Err(err) => Err(err),
+            },
+            None => Ok((None, gas_used)),
+        },
+        Err(err) => Err(err),
+    }
 }
 
 pub fn remove_encrypted_key(
     key: &[u8],
     context: &Ctx,
     contract_key: &ContractKey,
-) -> Result<(), DbError> {
+) -> Result<u64, WasmEngineError> {
     let scrambled_field_name = field_name_digest(key, contract_key);
 
     debug!("Removing scrambled field name: {:?}", scrambled_field_name);
 
     // Call remove_db (this bubbles up to Tendermint via ocalls and FFI to Go code)
     // fn remove_db(context: Ctx, key: &[u8]) {
-    remove_db(context, &scrambled_field_name).map_err(|err| {
+    let gas_used = remove_db(context, &scrambled_field_name).map_err(|err| {
         error!(
             "remove_db() got an error from ocall_remove_db, stopping wasm: {:?}",
             err
         );
-        DbError::FailedRemove
-    })
+        err
+    })?;
+    Ok(gas_used)
 }
 
 pub fn field_name_digest(field_name: &[u8], contract_key: &ContractKey) -> [u8; 32] {
-    let mut data: Vec<u8> = field_name.to_vec();
+    let mut data = field_name.to_vec();
     data.extend_from_slice(contract_key);
 
     sha_256(&data)
 }
 
 /// Safe wrapper around reads from the contract storage
-fn read_db(context: &Ctx, key: &[u8]) -> SgxResult<Option<Vec<u8>>> {
+fn read_db(context: &Ctx, key: &[u8]) -> Result<(Option<Vec<u8>>, u64), WasmEngineError> {
+    let mut ocall_return = OcallReturn::Success;
     let mut enclave_buffer = std::mem::MaybeUninit::<EnclaveBuffer>::uninit();
-    unsafe {
-        match imports::ocall_read_db(
-            enclave_buffer.as_mut_ptr(),
+    let mut vm_err = UntrustedVmError::null();
```

Comment:
Same as above, using `UntrustedVmError::default()` is probably cleaner.

Reply:
as i said, i liked `null` better, but i can change it to default :)